### PR TITLE
Imp: cache in reflection

### LINKED ISSUE #169: Use go reflection in JIT way
**What would you like to be added**:

Add cache for go reflection in this package.

**Why is this needed**:

To make this lib more efficiency.

### PR DESCRIPTION
**Pull request by micln**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #169 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by micln** [comment]
why cache in findField ?

image

$ go test -benchmem -run=^$ github.com/apache/dubbo-go-hessian2 -bench "^(BenchmarkDecode)$" -vet=off -v

before
```
BenchmarkDecode-8   	   61917	     18027 ns/op	    7448 B/op	     224 allocs/op
```

after
```
BenchmarkDecode-8   	   81508	     12387 ns/op	    7224 B/op	     126 allocs/op
```

File: go.mod [hunk 1]
```diff
@@ -1,5 +1,7 @@
 module github.com/apache/dubbo-go-hessian2
 
+go 1.12
```
Comment (wongoo) [inline go.mod:3]: remove the version limit

**Comment by wongoo** [comment]
great job

File: object.go [hunk 2]
```diff
@@ -164,21 +164,23 @@ func (e *Encoder) encObject(v POJO) error {
 	structs := []reflect.Value{vv}
 	for len(structs) > 0 {
 		vv := structs[0]
+		vvt := vv.Type()
 		num = vv.NumField()
 		for i = 0; i < num; i++ {
+			tf := vvt.Field(i)
```
Comment (micln) [inline object.go:170]: $ go test -benchmem -run=^$ github.com/apache/dubbo-go-hessian2 -bench "^(BenchmarkDecode)$" -vet=off -v

before

```
BenchmarkEncode-8   	   78986	     12693 ns/op	    3215 B/op	     122 allocs/op
```

after

```
BenchmarkEncode-8   	  105292	     10674 ns/op	    2974 B/op	      94 allocs/op
```

Comment (micln) [reply object.go:170]: simple but useful ... 😄

File: object.go [hunk 3]
```diff
@@ -292,36 +294,61 @@ func (d *Decoder) decClassDef() (interface{}, error) {
...
+	}
+
+	if len(indexes) > 0 {
+		return
+	}
+
+	defer func() {
+		typCache.(*sync.Map).Store(name, indexes)
+	}()
+
```
Comment (micln) [inline object.go:315]: should we do this in getclassInfo

File: pojo.go [hunk 4]
```diff
@@ -120,6 +120,11 @@ func RegisterPOJO(o POJO) int {
 	pojoRegistry.Lock()
 	defer pojoRegistry.Unlock()
 
+	if goName, ok := pojoRegistry.j2g[o.JavaClassName()]; ok {
+		return pojoRegistry.registry[goName].index
+	}
```
Comment (micln) [inline pojo.go:125]:
```
before:
BenchmarkEncode-8   	  105292	     10674 ns/op	    2974 B/op	      94 allocs/op
after:
BenchmarkEncode-8   	  211452	      5560 ns/op	    1771 B/op	      51 allocs/op
```

**Comment by micln** [comment]
`go test` doesn't work stable, this is a result seems average

As a whole, `hessian encode/decode` is performance 85% of `json.Marshal/Unmarshal` in cpu profiler now.

```
$ go test -benchmem -run=^$ github.com/apache/dubbo-go-hessian2 -bench "^B" -vet=off -v -count=5
goos: darwin
goarch: amd64
pkg: github.com/apache/dubbo-go-hessian2
BenchmarkJsonEncode
BenchmarkJsonEncode-8   	  249114	      4719 ns/op	     832 B/op	      15 allocs/op
BenchmarkJsonEncode-8   	  252224	      4862 ns/op	     832 B/op	      15 allocs/op
BenchmarkJsonEncode-8   	  240582	      4739 ns/op	     832 B/op	      15 allocs/op
BenchmarkJsonEncode-8   	  213283	      4784 ns/op	     832 B/op	      15 allocs/op
BenchmarkJsonEncode-8   	  227101	      4665 ns/op	     832 B/op	      15 allocs/op
BenchmarkEncode
BenchmarkEncode-8       	  182184	      5615 ns/op	    1771 B/op	      51 allocs/op
BenchmarkEncode-8       	  183007	      5565 ns/op	    1771 B/op	      51 allocs/op
BenchmarkEncode-8       	  218664	      5593 ns/op	    1771 B/op	      51 allocs/op
BenchmarkEncode-8       	  214704	      5886 ns/op	    1770 B/op	      51 allocs/op
BenchmarkEncode-8       	  181861	      5605 ns/op	    1770 B/op	      51 allocs/op
BenchmarkJsonDecode
BenchmarkJsonDecode-8   	  123667	      8412 ns/op	    1776 B/op	      51 allocs/op
BenchmarkJsonDecode-8   	  122796	      8497 ns/op	    1776 B/op	      51 allocs/op
BenchmarkJsonDecode-8   	  132103	      8471 ns/op	    1776 B/op	      51 allocs/op
BenchmarkJsonDecode-8   	  130687	      8492 ns/op	    1776 B/op	      51 allocs/op
BenchmarkJsonDecode-8   	  127668	      8476 ns/op	    1776 B/op	      51 allocs/op
BenchmarkDecode
BenchmarkDecode-8       	  107775	     10092 ns/op	    6424 B/op	      98 allocs/op
BenchmarkDecode-8       	  110996	      9950 ns/op	    6424 B/op	      98 allocs/op
BenchmarkDecode-8       	  111036	     10760 ns/op	    6424 B/op	      98 allocs/op
BenchmarkDecode-8       	  113151	     10063 ns/op	    6424 B/op	      98 allocs/op
BenchmarkDecode-8       	  109197	     10002 ns/op	    6424 B/op	      98 allocs/op
PASS
ok  	github.com/apache/dubbo-go-hessian2	28.680s
```

File: object.go [hunk 5]
```diff
@@ -337,13 +374,13 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
...
 
-		index, err := findField(fieldName, typ)
+		index, fieldStruct, err := findField(fieldName, typ)
 		if err != nil {
 			return nil, perrors.Errorf("can not find field %s", fieldName)
 		}
 
 		// skip unexported anonymous field
-		if vv.Type().FieldByIndex(index).PkgPath != "" {
+		if fieldStruct.PkgPath != "" {
```
Comment (micln) [inline object.go:383]: call rtype.FieldByIndex() repeatedly

File: object.go [hunk 6]
```diff
@@ -292,36 +295,72 @@ func (d *Decoder) decClassDef() (interface{}, error) {
...
 }
 
-func findField(name string, typ reflect.Type) ([]int, error) {
+type fieldInfo struct {
+	indexes []int
+	field   reflect.StructField
+}
+
+// map[rType][fieldName]indexes
+var _findFieldCache sync.Map
```
Comment (AlexStocks) [inline object.go:304]: pls change this ugly var name.

File: object_test.go [hunk 7]
```diff
@@ -659,3 +661,91 @@ func TestIssue150_EmbedStructJavaDecode(t *testing.T) {
...
+func (m Mix) JavaClassName() string {
+	return `test.mix`
+}
+
+func init() {
+	RegisterPOJO(new(Mix))
+}
+
+//
+// BenchmarkJsonEncode-8   	  217354	      4799 ns/op	     832 B/op	      15 allocs/op
```
Comment (gaoxinge) [inline object_test.go:684]: Placing benchmark informations into issue or readme may be better.

File: object.go [hunk 8]
```diff
@@ -292,36 +295,72 @@ func (d *Decoder) decClassDef() (interface{}, error) {
...
 
 		typField := typ.Field(i)
 
-		if val, has := typField.Tag.Lookup(tagIdentifier); has && strings.Compare(val, name) == 0 {
-			return []int{i}, nil
-		}
+		tagVal, hasTag := typField.Tag.Lookup(tagIdentifier)
 
 		fieldName := typField.Name
 		switch {
```
Comment (gaoxinge) [inline object.go:339]: Fixing `switch` to `if` may be better.

File: object.go [hunk 9]
```diff
@@ -292,36 +295,72 @@ func (d *Decoder) decClassDef() (interface{}, error) {
...
+	field   reflect.StructField
+}
+
+// map[rType][fieldName]indexes
+var _findFieldCache sync.Map
+
+func findField(name string, typ reflect.Type) (indexes []int, field reflect.StructField, err error) {
+	typCache, _ := _findFieldCache.Load(typ)
+	if typCache == nil {
+		typCache = &sync.Map{}
```
Comment (gaoxinge) [inline object.go:309]: Why use `&Sync.Map{}`, not just `Sync.Map{}`?

Comment (micln) [reply object.go:309]: Sync.Map is a structure which may lead some unnecessary copy

Comment (micln) [reply object.go:309]: Sync.Map is a structure which may lead some unnecessary copy

File: object_test.go (hunk 7 above)
Comment (micln) [reply object_test.go:684]: maybe it's convernients for next coders ?

**Comment by micln** [comment]
@AlexStocks @gaoxinge any problem ?

File: object_test.go (hunk 7 above)
Comment (gaoxinge) [reply object_test.go:684]: Ok.

File: object.go (hunk 9 above)
Comment (gaoxinge) [reply object.go:309]: Ok.

**Comment by gaoxinge** [comment]
@micln Good job. LGTM.

**Comment by wongoo** [comment]
@all pls resolve conversations if already resolved, so that make sure it can be merged.

**Comment by micln** [comment]
> @ALL pls resolve conversations if already resolved, so that make sure it can be merged.

done

File: object.go [hunk 10]
```diff
@@ -292,36 +295,71 @@ func (d *Decoder) decClassDef() (interface{}, error) {
...
-func findField(name string, typ reflect.Type) ([]int, error) {
+type fieldInfo struct {
+	indexes []int
+	field   reflect.StructField
+}
+
+// map[rType][fieldName]indexes
+var fieldIndexCache sync.Map
+
+func findField(name string, typ reflect.Type) (indexes []int, field reflect.StructField, err error) {
```
Comment (AlexStocks) [inline object.go:306]: in dubbo-go/dubbo-go-hessian2, we do not suggest set the var name for the return value. pls delete the return value name.

**Comment by AlexStocks** [comment]
LGTM

File: object.go [hunk 11]
```diff
@@ -375,11 +375,15 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
 	for i := 0; i < len(cls.fieldNameList); i++ {
 		fieldName := cls.fieldNameList[i]
 
-		index, fieldStruct, err := findField(fieldName, typ)
+		index, fieldStruct, err := findFieldWithCache(fieldName, typ)
 		if err != nil {
 			return nil, perrors.Errorf("can not find field %s", fieldName)
 		}
 
+		if fieldStruct == nil {
```
Comment (wongoo) [inline object.go:383]: the fieldStruct won't be nil if err is nil, this check is redundant.  And should it return error when fieldStruct is nil?

Comment (micln) [reply object.go:383]: because of #183, I avoid to depend on
 err.

**Pull request merged**